lib/gli/exceptions.rb

```diff
@@ -59,17 +59,48 @@ def exit_code
     end
   end
 
-  class MissingRequiredArgumentsException < BadCommandLine
+  class BadCommandOptionsOrArguments < BadCommandLine
     # The command for which the argument was unknown
     attr_reader :command_in_context
-    # +message+:: the error message to show the user
-    # +command+:: the command we were using to parse command-specific options
     def initialize(message,command)
       super(message)
       @command_in_context = command
     end
   end
 
+  class MissingRequiredArgumentsException < BadCommandOptionsOrArguments
+    # +message+:: the error message to show the user
+    # +command+:: the command we were using to parse command-specific options
+    def initialize(message,command)
+      super(message,command)
+    end
+  end
+
+  class TooFewArgumentsException < MissingRequiredArgumentsException
+    def initialize(command,num_arguments_received,min_number_of_arguments)
+      message = "#{command.name} expected at least #{min_number_of_arguments} arguments, but was given only #{num_arguments_received}"
+      super(message,command)
+    end
+  end
+  class TooManyArgumentsException < MissingRequiredArgumentsException
+    def initialize(command,num_arguments_received,max_number_of_arguments)
+      message = if max_number_of_arguments == 0
+                  "#{command.name} expected no arguments, but received #{num_arguments_received}"
+                else
+                  "#{command.name} expected only #{max_number_of_arguments} arguments, but received #{num_arguments_received}"
+                end
+      super(message,command)
+    end
+  end
```

Comment:
Somewhat minor design thing, but it feels like this be one exception class that does the work of determining the message from the input passed to it, which would reduce the amount of inheritance going on.

Something like:

```ruby
  class BadArgumentsException < BadCommandOptionsOrArguments
    def initialize(command,args_received,min,max)
      message = if args_received < min
        "#{command.name} expected at least #{min} arguments, but was given only #{args_received}"
      else 
        if max == 0
          "#{command.name} expected no arguments, but received #{args_received}"
        else
          "#{command.name} expected only #{max} arguments, but received #{args_received}"
       end
      end

      super(message,command)
    end
  end
```

and the caller would become something like:

```ruby
 if (arguments.size < min) || (arguments.size > max)
   raise BadArgumentsException.new(command,arguments.size,min,max)
end
```

Reply:
I did try this initially and ended up where I did for no particular reason.  Let me try yours out as it does seem nice to have fewer classes